server/src/test/java/org/elasticsearch/action/ActionListenerTests.java

```diff
@@ -370,6 +374,72 @@ public void onFailure(Exception e) {
         assertThat(exReference.get(), instanceOf(IllegalArgumentException.class));
     }
 
+    public void testAssertAtLeastOnceWillLogAssertionErrorWhenNotResolved() throws Exception {
+        try (MockLog mockLog = MockLog.capture(ExceptionsHelper.class)) {
+            mockLog.addExpectation(new MockLog.PatternSeenEventExpectation(
+                "action listener not called",
+                ExceptionsHelper.class.getName(),
+                Level.ERROR,
+                ".*Expected listener NoopActionListener to be called at least once, " +
+                    "but it was never called\\. Created\\:(.*|\\s)*"
+                ));
+            final AtomicReference<ActionListener<Object>> listenerRef = new AtomicReference<>(
+                ActionListener.assertAtLeastOnce(ActionListener.noop()));
+            assertBusy(() -> {
+                // Nullify reference so it becomes unreachable
+                listenerRef.set(null);
+                System.gc();
+                mockLog.assertAllExpectationsMatched();
+            });
+        }
+    }
+
+    public void testAssertAtLeastOnceWillInvokeListenerWhenNotResolved() throws Exception {
+        AtomicReference<Exception> listenerCreatedAt = new AtomicReference<>();
+        AtomicReference<ActionListener<?>> notCalledListener = new AtomicReference<>();
+        AtomicReference<ActionListener<Object>> listenerRef = new AtomicReference<>(
+            new ActionListenerImplementations.AssertAtLeastOnceActionListener<>(
+                ActionListener.noop(),
+                (listener, createdAt) -> {
+                    logger.info("NotCalledListener called!");
+                    notCalledListener.set(listener);
+                    listenerCreatedAt.set(createdAt);
+                }));
+        assertBusy(() -> {
+            // Nullify reference so it becomes unreachable
+            listenerRef.set(null);
+            System.gc();
+            assertNotNull(notCalledListener.get());
+            assertNotNull(listenerCreatedAt.get());
+        });
+    }
```

Comment:
This and the next one test the `ActionListenerImplementations.AssertAtLeastOnceActionListener` directly, as it was easier to do with a specific `NotCalledListener`. Using `MockLog#assertAllExpectationsMatched` seemed brittle when testing that a log message didn't appear.